File: DotNetPolyConsole/ColorConverter.vb
Public Class ColorConverter

    Private Shared MatchingCache As New Collections.Generic.Dictionary(Of System.Drawing.Color, ConsoleColor)

    Friend Shared Function ClosestConsoleColor(color As System.drawing.color) As ConsoleColor
        Dim ret As ConsoleColor

        If MatchingCache.TryGetValue(color, ret) Then
            Return ret
        End If

        Dim rr As Double = color.R
        Dim gg As Double = color.G
        Dim bb As Double = color.B
        Dim delta = Double.MaxValue

        For Each cc As ConsoleColor In [Enum].GetValues(GetType(ConsoleColor))
            Dim name = [Enum].GetName(GetType(ConsoleColor), cc)
            Dim c = System.Drawing.Color.FromName(CStr(Microsoft.VisualBasic.IIf(name = "DarkYellow", "Orange", name)))
            Dim t = Math.Pow(c.R - rr, 2) + Math.Pow(c.G - gg, 2) + Math.Pow(c.B - bb, 2)
            If t = 0 Then
                If Not MatchingCache.ContainsKey(color) Then MatchingCache.Add(color, cc)
                Return cc
            End If
            If t < delta Then
                delta = t
                ret = cc
            End If
        Next
        If Not MatchingCache.ContainsKey(color) Then MatchingCache.Add(color, ret)
        Return ret
    End Function

End Class

File: DotNetPolyConsole/ColorConverter.cs
using System;

namespace DotNetPolyConsole
{
    public class ColorConverter
    {
        private static System.Collections.Generic.Dictionary<System.Drawing.Color, ConsoleColor> MatchingCache = new System.Collections.Generic.Dictionary<System.Drawing.Color, ConsoleColor>();

        internal static ConsoleColor ClosestConsoleColor(System.Drawing.Color color)
        {
            ConsoleColor ret;

            if (MatchingCache.TryGetValue(color, out ret))
                return ret;

            double rr = color.R;
            double gg = color.G;
            double bb = color.B;
            var delta = double.MaxValue;

            foreach (ConsoleColor cc in Enum.GetValues(typeof(ConsoleColor)))
            {
                var name = Enum.GetName(typeof(ConsoleColor), cc);
                var c = System.Drawing.Color.FromName(System.Convert.ToString(name == "DarkYellow" ? "Orange" : name));
                var t = Math.Pow(c.R - rr, 2) + Math.Pow(c.G - gg, 2) + Math.Pow(c.B - bb, 2);
                if (t == 0)
                {
                    if (!MatchingCache.ContainsKey(color))
                        MatchingCache.Add(color, cc);
                    return cc;
                }
                if (t < delta)
                {
                    delta = t;
                    ret = cc;
                }
            }
            if (!MatchingCache.ContainsKey(color))
                MatchingCache.Add(color, ret);
            return ret;
        }
    }
}
